lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,135 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Dict, Iterator, List, MutableMapping, Set, Tuple, Union
+
+from streamlit.errors import StreamlitAPIException
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+
+@dataclass
+class QueryParams(MutableMapping[str, str]):
+    """A lightweight wrapper of a dict that sends forwardMsgs when state changes.
+    It stores str keys with str and List[str] values.
+    """
+
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __iter__(self) -> Iterator[str]:
+        self._ensure_single_query_api_used()
+        return iter(self._query_params.keys())
+
+    def __getitem__(self, key: str) -> str:
+        """Retrieves a value for a given key in query parameters.
+        Returns the last item in a list or an empty string if empty.
+        If the key is not present, raise KeyError.
+        """
+        self._ensure_single_query_api_used()
+        try:
+            value = self._query_params[key]
+            if isinstance(value, list):
+                if len(value) == 0:
+                    return ""
+                else:
+                    # Return the last value to mimic Tornado's behavior
+                    # https://www.tornadoweb.org/en/stable/web.html#tornado.web.RequestHandler.get_query_argument
+                    return value[-1]
+            return value
+        except KeyError:
+            raise KeyError(missing_key_error_message(key))
+
+    def __setitem__(
+        self, key: str, value: Union[str, Set[str], Tuple[str], List[str]]
```

Comment:
nit: Why not allow any `Iterable` here? e.g.:

```suggestion
        self, key: str, value: Union[str,  Iterable[str]]
```

and below we can then just check via:

```
from collections.abc import Iterable


if isinstance(value, Iterable):
            self._query_params[key] = [str(item) for item in value]
```